src/ray/core_worker/transport/direct_task_transport.h

```diff
@@ -177,25 +202,32 @@ class CoreWorkerDirectTaskSubmitter {
   /// (1) The lease client through which the worker should be returned
   /// (2) The expiration time of a worker's lease.
   /// (3) The number of tasks that are currently in flight to the worker
+  /// (4) The current scheduling key of the tasks submitted to the worker
+  /// (5) The set of TaskIDs of the tasks that can be stolen by other workers
   struct LeaseEntry {
     std::shared_ptr<WorkerLeaseInterface> lease_client_;
     int64_t lease_expiration_time_;
     uint32_t tasks_in_flight_;
+    SchedulingKey current_scheduling_key_;
+    absl::flat_hash_set<TaskID> stealable_tasks_;
```

Comment:
The reason why we need to keep track of the number of stealable tasks as well as the number of tasks in flight (the former are a subset of the latter) is that the two quantities are often not equal, and using only one variable instead of two would cause some complications, inefficiencies, and perhaps errors.

When we steal a bunch of tasks from a worker, assuming that the reply to the thief's `StealWork` RPC arrives to the owner before the reply to the victim's `PushNormalTask` RPCs for the stolen tasks, we remove the stolen tasks from the `stealable_tasks_` hash_set, while the victim's `tasks_in_flight_` value initially remains unchanged. This reflects the fact that the stolen tasks are still technically in flight to the victim. Knowing how many tasks are in flight to a worker (even if they have been stolen) is important, for example, to know when to return the worker to the Raylet. So even if the stolen tasks are not going to be executed by the victim anymore, it is still useful for us to know that they are still "in flight", in the sense that their respective RPCs have not completed yet.

On the other hand, knowing the number of tasks in flight does not allow us, on its own, to perform work stealing in an efficient way. If we rely on it to choose which worker to steal from (picking the one with more tasks in flight), we will frequently make a suboptimal choice, as many times the number of tasks up for grab will be lower than the reported number of tasks in flight. 

Having decided to simultaneously keep track of the number of stealable tasks, as well as the number of tasks in flight to a worker, why use a `flat_hash_set` containing the TaskIDs of the `stealable_tasks_`, instead of just using an integer? The reason for this is to avoid a race condition between a thief's `StealWork` RPC callback and the victim's `PushNormalTask` RPC callbacks. When the thief steals some tasks from a victim, the reply to the `StealWork` RPC usually comes back to the owner earlier than the replies to the PushNormalTask RPCs. However, this might not always be the case. If the PushNormalTask callbacks are run first, they will be the ones responsible for decrementing the number of stealable tasks (to maintain the invariant that `stealable_tasks_.size() <= tasks_in_flight`). The challenge is that we must decrement the number of stealable tasks exactly once for each stolen task, and decrementing an integer is not an idempotent operation. On the other hand, deleting a `TaskID` from a set is idempotent, so we don't have to worry about finding a way to let the `PushNormalTask` and the `StealWork` callbacks coordinate to decrement the counter only once.